### PR TITLE
add convenience methods for standard events

### PR DESCRIPTION
**Pull request by yjbanov**
This is still WIP as there are many more standard event types. I would like to check first that this is what we want our convenience APIs to look like.

### CODE REVIEW

**Comment by yjbanov** [comment]
> I wonder if we should let the user set custom parameters when using the helpers, or if we should make them use logEvent

Good point. I'll check if Firebase supports custom properties on standard events.

> I assume you know there are some helpers that you haven't gotten around to implementing yet.

Yep, this is WIP. I will continue filling in the missing events tomorrow. Just wanted to check with you lest I veer off in the wrong direction.

**Comment by yjbanov** [comment]
Added more convenience APIs. I haven't found evidence that Firebase accepts custom parameters for standard events. We can add them later in a non-breaking manner, so I'd prefer that we postpone it until we know for sure (removing API is harder than adding it). PTAL. I'll add some tests too.

File: lib/firebase_analytics.dart [hunk 1]
```diff
@@ -17,12 +21,19 @@ class FirebaseAnalytics {
...
     if (_reservedEventNames.contains(name)) {
       throw new ArgumentError.value(name, 'name',
           'Event name is reserved and cannot be used');
     }
 
+    const String kReservedPrefix = 'firebase_';
+
+    if (name.startsWith(kReservedPrefix)) {
+      throw new ArgumentError.value(name, 'name',
+          'Previx "$kReservedPrefix" is reserved and cannot be used.');
```
Comment (collinjackson) [inline lib/firebase_analytics.dart]: s/Previx/Prefix/

File: lib/firebase_analytics.dart [hunk 2]
```diff
@@ -96,22 +405,303 @@ class FirebaseAnalytics {
...
 /// Mutates the [parameters] map, removing all parameters whose value is `null`.
-Map<String, String> _filterOutNulls(Map<String, String> parameters) {
+Map<String, dynamic> _filterOutNulls(Map<String, dynamic> parameters) {
   // To prevent concurrent modification
   final List<String> copyOfKeys = parameters.keys.toList();
   for (String key in copyOfKeys) {
     if (parameters[key] == null)
       parameters.remove(key);
   }
+  return parameters;
```
Comment (collinjackson) [inline lib/firebase_analytics.dart]: The concurrency comment takes a bit of effort to understand, and it seems to me that constructing a new Map would be safer / more readable than deletion.

```
    return new Map<String, dynamic>.fromIterable(
      parameters.keys.where((String key) => parameters[key] != null),
      value: (String key) => parameters[key],
    );
```

File: lib/firebase_analytics.dart (hunk 1 above)
Comment (yjbanov) [inline lib/firebase_analytics.dart]: Done.

File: lib/firebase_analytics.dart (hunk 2 above)
Comment (yjbanov) [inline lib/firebase_analytics.dart]: Done.

**Comment by collinjackson** [comment]
lgtm

File: lib/firebase_analytics.dart [hunk 3]
```diff
@@ -678,15 +678,15 @@ class FirebaseAnalytics {
...
 Map<String, dynamic> _filterOutNulls(Map<String, dynamic> parameters) {
-  // To prevent concurrent modification
-  final List<String> copyOfKeys = parameters.keys.toList();
-  for (String key in copyOfKeys) {
-    if (parameters[key] == null)
-      parameters.remove(key);
-  }
-  return parameters;
+  final Map<String, dynamic> filtered = <String, dynamic>{};
+  filtered.forEach((String key, dynamic value) {
```
Comment (collinjackson) [inline lib/firebase_analytics.dart]: I think you meant parameters instead of filtered here? Or you could use the fromIterable version I suggested

Comment (yjbanov) [inline lib/firebase_analytics.dart]: Good catch. Fixed. Also added a bunch of tests. PTAL.

**Comment by collinjackson** [comment]
lgtm thanks!

**Pull request merged**